Fix intermittent test failures in test_steps_with_retry

## overview

This testing code fails intermittently. CI/CD testing could should pass deterministically.

```
____________________________ test_steps_with_retry _____________________________

durable_runner = <test.conftest.TestRunnerAdapter object at 0x7ff1ee7f7930>

    @pytest.mark.example
    @pytest.mark.durable_execution(
        handler=steps_with_retry.handler,
        lambda_function_name="steps with retry",
    )
    def test_steps_with_retry(durable_runner):
        """Test steps_with_retry pattern."""
        with durable_runner:
>           result = durable_runner.run(input={"name": "test-item"}, timeout=30)
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

examples/test/test_steps_with_retry.py:19: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
examples/test/conftest.py:106: in run
    return self._runner.run(input=input, timeout=timeout)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <aws_durable_execution_sdk_python_testing.runner.DurableFunctionTestRunner object at 0x7ff1ee7f6c30>
input = {'name': 'test-item'}, timeout = 30, function_name = 'test-function'
execution_name = 'execution-name', account_id = '123456789012'

    def run(
        self,
        input: str | None = None,  # noqa: A002
        timeout: int = 900,
        function_name: str = "test-function",
        execution_name: str = "execution-name",
        account_id: str = "123456789012",
    ) -> DurableFunctionTestResult:
        start_input = StartDurableExecutionInput(
            account_id=account_id,
            function_name=function_name,
            function_qualifier="$LATEST",
            execution_name=execution_name,
            execution_timeout_seconds=timeout,
            execution_retention_period_days=7,
            invocation_id="inv-12345678-1234-1234-1234-123456789012",
            trace_fields={"trace_id": "abc123", "span_id": "def456"},
            tenant_id="tenant-001",
            input=input,
        )
    
        output: StartDurableExecutionOutput = self._executor.start_execution(
            start_input
        )
    
        if output.execution_arn is None:
            msg_arn: str = "Execution ARN must exist to run test."
            raise DurableFunctionsTestError(msg_arn)
    
        # Block until completion
        completed = self._executor.wait_until_complete(output.execution_arn, timeout)
    
        if not completed:
            msg_timeout: str = "Execution did not complete within timeout"
    
>           raise TimeoutError(msg_timeout)
E           TimeoutError: Execution did not complete within timeout

src/aws_durable_execution_sdk_python_testing/runner.py:553: TimeoutError
------------------------------ Captured log call -------------------------------
ERROR    aws_durable_execution_sdk_python.operation.step:step.py:174 ❌ failed step for id: 1ced8f5be2db23a6513eba4d819c73806424748a7bc6fa0d792cc1c7d1775a97, name: get_item_poll_1
Traceback (most recent call last):
  File "/home/runner/.local/share/hatch/env/virtual/aws-durable-execution-sdk-python-testing/6OKiTmoA/test/lib/python3.14/site-packages/aws_durable_execution_sdk_python/operation/step.py", line 139, in step_handler
    raw_result: T = func(step_context)
                    ~~~~^^^^^^^^^^^^^^
  File "/home/runner/work/aws-durable-execution-sdk-python-testing/aws-durable-execution-sdk-python-testing/examples/src/steps_with_retry.py", line 52, in <lambda>
    lambda _, n=name: simulated_get_item(n),
                      ~~~~~~~~~~~~~~~~~~^^^
  File "/home/runner/work/aws-durable-execution-sdk-python-testing/aws-durable-execution-sdk-python-testing/examples/src/steps_with_retry.py", line 20, in simulated_get_item
    raise RuntimeError(msg)
RuntimeError: Random failure
ERROR    aws_durable_execution_sdk_python.operation.step:step.py:174 ❌ failed step for id: 1ced8f5be2db23a6513eba4d819c73806424748a7bc6fa0d792cc1c7d1775a97, name: get_item_poll_1
Traceback (most recent call last):
  File "/home/runner/.local/share/hatch/env/virtual/aws-durable-execution-sdk-python-testing/6OKiTmoA/test/lib/python3.14/site-packages/aws_durable_execution_sdk_python/operation/step.py", line 139, in step_handler
    raw_result: T = func(step_context)
                    ~~~~^^^^^^^^^^^^^^
  File "/home/runner/work/aws-durable-execution-sdk-python-testing/aws-durable-execution-sdk-python-testing/examples/src/steps_with_retry.py", line 52, in <lambda>
    lambda _, n=name: simulated_get_item(n),
                      ~~~~~~~~~~~~~~~~~~^^^
  File "/home/runner/work/aws-durable-execution-sdk-python-testing/aws-durable-execution-sdk-python-testing/examples/src/steps_with_retry.py", line 20, in simulated_get_item
    raise RuntimeError(msg)
RuntimeError: Random failure
ERROR    aws_durable_execution_sdk_python.operation.step:step.py:174 ❌ failed step for id: 1ced8f5be2db23a6513eba4d819c73806424748a7bc6fa0d792cc1c7d1775a97, name: get_item_poll_1
Traceback (most recent call last):
  File "/home/runner/.local/share/hatch/env/virtual/aws-durable-execution-sdk-python-testing/6OKiTmoA/test/lib/python3.14/site-packages/aws_durable_execution_sdk_python/operation/step.py", line 139, in step_handler
    raw_result: T = func(step_context)
                    ~~~~^^^^^^^^^^^^^^
  File "/home/runner/work/aws-durable-execution-sdk-python-testing/aws-durable-execution-sdk-python-testing/examples/src/steps_with_retry.py", line 52, in <lambda>
    lambda _, n=name: simulated_get_item(n),
                      ~~~~~~~~~~~~~~~~~~^^^
  File "/home/runner/work/aws-durable-execution-sdk-python-testing/aws-durable-execution-sdk-python-testing/examples/src/steps_with_retry.py", line 20, in simulated_get_item
    raise RuntimeError(msg)
RuntimeError: Random failure
=============================== warnings summary ===============================
examples/test/test_block_example.py: 3 warnings
examples/test/test_callback.py: 1 warning
examples/test/test_callback_permutations.py: 1 warning
examples/test/test_hello_world.py: 1 warning
examples/test/test_logger_example.py: 1 warning
examples/test/test_map_operations.py: 1 warning
examples/test/test_parallel.py: 1 warning
examples/test/test_run_in_child_context.py: 1 warning
examples/test/test_step.py: 1 warning
examples/test/test_step_permutations.py: 3 warnings
examples/test/test_step_semantics_at_most_once.py: 1 warning
examples/test/test_step_with_retry.py: 1 warning
examples/test/test_steps_with_retry.py: 5 warnings
examples/test/test_wait.py: 3 warnings
examples/test/test_wait_for_condition.py: 5 warnings
examples/test/test_wait_permutations.py: 3 warnings
  /home/runner/work/aws-durable-execution-sdk-python-testing/aws-durable-execution-sdk-python-testing/src/aws_durable_execution_sdk_python_testing/scheduler.py:173: DeprecationWarning: 'asyncio.iscoroutinefunction' is deprecated and slated for removal in Python 3.16; use inspect.iscoroutinefunction() instead
    if asyncio.iscoroutinefunction(func):

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=========================== short test summary info ============================
FAILED examples/test/test_steps_with_retry.py::test_steps_with_retry - TimeoutError: Execution did not complete within timeout
================== 1 failed, 17 passed, 32 warnings in 45.66s ==================
Error: Process completed with exit code 1.
```


## ac
1. Amend test code either to pass deterministically. If the purpose is to test retries, the code should retry deterministically and assert on the appropriate success or failure condition to exit without failing intermittently.

--- a/examples/src/step/step_with_retry.py
+++ b/examples/src/step/step_with_retry.py
@@ -1,4 +1,4 @@
-from random import random
+from itertools import count
 from typing import Any
 
 from aws_durable_execution_sdk_python.config import StepConfig
@@ -14,13 +14,19 @@ from aws_durable_execution_sdk_python.retries import (
 )
 
 
+# Counter for deterministic behavior across retries
+_attempts = count(1)  # starts from 1
+
+
 @durable_step
 def unreliable_operation(
     _step_context: StepContext,
 ) -> str:
-    failure_threshold = 0.5
-    if random() > failure_threshold:  # noqa: S311
-        msg = "Random error occurred"
+    # Use counter for deterministic behavior
+    # Will fail on first attempt, succeed on second
+    attempt = next(_attempts)
+    if attempt < 2:
+        msg = f"Attempt {attempt} failed"
         raise RuntimeError(msg)
     return "Operation succeeded"
 

--- a/examples/src/step/steps_with_retry.py
+++ b/examples/src/step/steps_with_retry.py
@@ -1,10 +1,10 @@
 """Example demonstrating multiple steps with retry logic."""
 
-from random import random
+from itertools import count
 from typing import Any
 
-from aws_durable_execution_sdk_python.config import StepConfig, Duration
-from aws_durable_execution_sdk_python.context import DurableContext
+from aws_durable_execution_sdk_python.config import Duration, StepConfig
+from aws_durable_execution_sdk_python.context import DurableContext, StepContext
 from aws_durable_execution_sdk_python.execution import durable_execution
 from aws_durable_execution_sdk_python.retries import (
     RetryStrategyConfig,
@@ -12,18 +12,26 @@ from aws_durable_execution_sdk_python.retries import (
 )
 
 
-def simulated_get_item(name: str) -> dict[str, Any] | None:
-    """Simulate getting an item that may fail randomly."""
-    # Fail 50% of the time
-    if random() < 0.5:  # noqa: S311
+# Counter for deterministic behavior across retries
+_attempts = count(1)  # starts from 1
+
+
+def simulated_get_item(_step_context: StepContext, name: str) -> dict[str, Any] | None:
+    """Simulate getting an item with deterministic counter-based behavior."""
+    # Use counter for deterministic behavior
+    attempt = next(_attempts)
+
+    # Fail on first attempt
+    if attempt == 1:
         msg = "Random failure"
         raise RuntimeError(msg)
 
-    # Simulate finding item after some attempts
-    if random() > 0.3:  # noqa: S311
-        return {"id": name, "data": "item data"}
+    # Return None on second attempt (poll 1)
+    if attempt == 2:
+        return None
 
-    return None
+    # Return item on third attempt (poll 2, after retry)
+    return {"id": name, "data": "item data"}
 
 
 @durable_execution
@@ -49,7 +57,7 @@ def handler(event: Any, context: DurableContext) -> dict[str, Any]:
 
             # Try to get the item with retry
             get_response = context.step(
-                lambda _, n=name: simulated_get_item(n),
+                lambda _, n=name: simulated_get_item(_, n),
                 name=f"get_item_poll_{poll_count}",
                 config=step_config,
             )

--- a/examples/test/step/test_step_with_retry.py
+++ b/examples/test/step/test_step_with_retry.py
@@ -3,7 +3,6 @@
 import pytest
 from aws_durable_execution_sdk_python.execution import InvocationStatus
 from aws_durable_execution_sdk_python.lambda_service import OperationType
-
 from src.step import step_with_retry
 from test.conftest import deserialize_operation_payload
 
@@ -14,20 +13,28 @@ from test.conftest import deserialize_operation_payload
     lambda_function_name="step with retry",
 )
 def test_step_with_retry(durable_runner):
-    """Test step with retry configuration."""
+    """Test step with retry configuration.
+
+    With counter-based deterministic behavior:
+    - Attempt 1: counter = 1 < 2 → raises RuntimeError ❌
+    - Attempt 2: counter = 2 >= 2 → succeeds ✓
+
+    The function deterministically fails once then succeeds on the second attempt.
+    """
     with durable_runner:
         result = durable_runner.run(input="test", timeout=30)
 
-    # The function uses random() so it may succeed or fail
-    # We just verify it completes and has retry configuration
-    assert result.status in [InvocationStatus.SUCCEEDED, InvocationStatus.FAILED]
+    # With counter-based deterministic behavior, succeeds on attempt 2
+    assert result.status is InvocationStatus.SUCCEEDED
+    assert deserialize_operation_payload(result.result) == "Operation succeeded"
 
-    # Verify step operation exists
+    # Verify step operation exists with retry details
     step_ops = [
         op for op in result.operations if op.operation_type == OperationType.STEP
     ]
-    assert len(step_ops) >= 1
+    assert len(step_ops) == 1
 
-    # If it succeeded, verify the result
-    if result.status is InvocationStatus.SUCCEEDED:
-        assert deserialize_operation_payload(result.result) == "Operation succeeded"
+    # The step should have succeeded on attempt 2 (after 1 failure)
+    # Attempt numbering: 1 (initial attempt), 2 (first retry)
+    step_op = step_ops[0]
+    assert step_op.attempt == 2  # Succeeded on first retry (1-indexed: 2=first retry)

--- a/examples/test/step/test_steps_with_retry.py
+++ b/examples/test/step/test_steps_with_retry.py
@@ -3,7 +3,6 @@
 import pytest
 from aws_durable_execution_sdk_python.execution import InvocationStatus
 from aws_durable_execution_sdk_python.lambda_service import OperationType
-
 from src.step import steps_with_retry
 from test.conftest import deserialize_operation_payload
 
@@ -14,20 +13,40 @@ from test.conftest import deserialize_operation_payload
     lambda_function_name="steps with retry",
 )
 def test_steps_with_retry(durable_runner):
-    """Test steps_with_retry pattern."""
+    """Test steps_with_retry pattern.
+
+    With counter-based deterministic behavior:
+    - Poll 1, Attempt 1: counter = 1 → raises RuntimeError ❌
+    - Poll 1, Attempt 2: counter = 2 → returns None
+    - Poll 2, Attempt 1: counter = 3 → returns item ✓
+
+    The function finds the item on poll 2 after 1 retry on poll 1.
+    """
     with durable_runner:
         result = durable_runner.run(input={"name": "test-item"}, timeout=30)
 
     assert result.status is InvocationStatus.SUCCEEDED
 
-    # Result should be either success with item or error
-    assert isinstance(deserialize_operation_payload(result.result), dict)
-    assert "success" in deserialize_operation_payload(
-        result.result
-    ) or "error" in deserialize_operation_payload(result.result)
+    # With counter-based deterministic behavior, finds item on poll 2
+    result_data = deserialize_operation_payload(result.result)
+    assert isinstance(result_data, dict)
+    assert result_data.get("success") is True
+    assert result_data.get("pollsRequired") == 2
+    assert "item" in result_data
+    assert result_data["item"]["id"] == "test-item"
 
-    # Verify step operations exist (polling steps)
+    # Verify step operations exist
     step_ops = [
         op for op in result.operations if op.operation_type == OperationType.STEP
     ]
-    assert len(step_ops) >= 1
+    # Should have exactly 2 step operations (poll 1 and poll 2)
+    assert len(step_ops) == 2
+
+    # Poll 1: succeeded after 1 retry (returned None)
+    assert step_ops[0].name == "get_item_poll_1"
+    assert step_ops[0].result == "null"
+    assert step_ops[0].attempt == 2  # 1 retry occurred (1-indexed: 2=first retry)
+
+    # Poll 2: succeeded immediately (returned item)
+    assert step_ops[1].name == "get_item_poll_2"
+    assert step_ops[1].attempt == 1  # No retries needed (1-indexed: 1=initial)

--- a/src/aws_durable_execution_sdk_python_testing/checkpoint/processors/base.py
+++ b/src/aws_durable_execution_sdk_python_testing/checkpoint/processors/base.py
@@ -12,6 +12,7 @@ from aws_durable_execution_sdk_python.lambda_service import (
     ContextDetails,
     ExecutionDetails,
     Operation,
+    OperationAction,
     OperationStatus,
     OperationType,
     OperationUpdate,
@@ -72,9 +73,15 @@ class OperationProcessor:
         )
 
     def _create_step_details(
-        self, update: OperationUpdate, current_operation: Operation | None = None
+        self,
+        update: OperationUpdate,
+        current_operation: Operation | None = None,
     ) -> StepDetails | None:
-        """Create StepDetails from OperationUpdate."""
+        """Create StepDetails from OperationUpdate.
+
+        Automatically increments attempt count for RETRY, SUCCEED, and FAIL actions.
+        """
+
         attempt: int = 0
         next_attempt_timestamp: datetime.datetime | None = None
 
@@ -84,6 +91,13 @@ class OperationProcessor:
                 next_attempt_timestamp = (
                     current_operation.step_details.next_attempt_timestamp
                 )
+            # Increment attempt for RETRY, SUCCEED, and FAIL actions
+            if update.action in {
+                OperationAction.RETRY,
+                OperationAction.SUCCEED,
+                OperationAction.FAIL,
+            }:
+                attempt += 1
             return StepDetails(
                 attempt=attempt,
                 next_attempt_timestamp=next_attempt_timestamp,

--- a/tests/checkpoint/processors/step_test.py
+++ b/tests/checkpoint/processors/step_test.py
@@ -230,6 +230,7 @@ def test_process_succeed_action_with_current_operation():
 
     current_op = Mock()
     current_op.start_timestamp = datetime.now(UTC)
+    current_op.step_details = StepDetails()
 
     update = OperationUpdate(
         operation_id="step-123",
@@ -243,6 +244,7 @@ def test_process_succeed_action_with_current_operation():
 
     assert result.start_timestamp == current_op.start_timestamp
     assert result.status == OperationStatus.SUCCEEDED
+    assert result.step_details.attempt == 1
 
 
 def test_process_fail_action():
@@ -274,6 +276,7 @@ def test_process_fail_action_with_current_operation():
 
     current_op = Mock()
     current_op.start_timestamp = datetime.now(UTC)
+    current_op.step_details = StepDetails()
 
     error = ErrorObject.from_message("step failed")
     update = OperationUpdate(
@@ -288,6 +291,7 @@ def test_process_fail_action_with_current_operation():
 
     assert result.start_timestamp == current_op.start_timestamp
     assert result.status == OperationStatus.FAILED
+    assert result.step_details.attempt == 1
 
 
 def test_process_invalid_action():

